### PR TITLE
Fixing Google Maps markers and appointments

### PR DESCRIPTION
**Pull request by htysc**
And Appointments visualization

### CODE REVIEW

File: frontend/src/components/Providers.js [hunk 1]
```diff
@@ -86,16 +84,16 @@ export default function Providers() {
         for (const occ of innerData.characteristicOccurrences) {
           if (occ.occurrenceOf?.name === 'Organization Name') {
             providerData.name = occ.dataStringValue;
-          } else if (occ.occurrenceOf?.name === 'Organization Address') {
-            providerData.address = occ.objectValue;
```
Comment (htysc) [inline frontend/src/components/Providers.js:90]: Is the Organization Address required?

Comment (LesterLyu) [reply frontend/src/components/Providers.js:90]: It is optional, the provider can be either an organization or a volunteer.

Comment (htysc) [reply frontend/src/components/Providers.js:90]: Do we need to have it? There is a separate address field for both organizations and volunteers.

File: frontend/src/components/Providers.js [hunk 2]
```diff
@@ -86,16 +84,16 @@ export default function Providers() {
...
           } else if (occ.occurrenceOf?.name === 'Email') {
             providerData.email = occ.dataStringValue;
           } else if (occ.occurrenceOf?.name === 'First Name') {
             providerData.firstName = occ.dataStringValue;
           } else if (occ.occurrenceOf?.name === 'Last Name') {
             providerData.lastName = occ.dataStringValue;
-          }
-
+          } else if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleProvider(provider.type, provider._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/Providers.js:94]: This is not efficient, please make the backend populate this field.

File: frontend/src/components/Providers.js (hunk 1 above)
Comment (LesterLyu) [reply frontend/src/components/Providers.js:90]: This can be removed.

File: frontend/src/components/Providers.js (hunk 2 above)
Comment (LesterLyu) [reply frontend/src/components/Providers.js:94]: For some reason Service Provider is not part of the generic data. Logics are under [here](https://github.com/csse-uoft/SNM-I/blob/master/backend/services/genericData/serviceProvider.js).
Add `organization.address` to the [populates array](https://github.com/csse-uoft/SNM-I/blob/master/backend/services/genericData/serviceProvider.js#L29) will make it populate the `address` field for the frontend.

File: frontend/src/components/needOccurrence/NeedOccurrences.js [hunk 3]
```diff
@@ -87,6 +79,13 @@ export default function NeedOccurrences() {
           _id: needOccurrence.occurrenceOf._id,
         }
       }
+      if (needOccurrence.characteristicOccurrences)
+        for (const occ of needOccurrence.characteristicOccurrences) {
+          if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleGeneric("needOccurrence", needOccurrence._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/needOccurrence/NeedOccurrences.js:85]: Same as above.

File: frontend/src/components/outcomeOccurrence/OutcomeOccurrences.js [hunk 4]
```diff
@@ -99,6 +91,13 @@ export default function OutcomeOccurrences() {
           _id: outcomeOccurrence.occurrenceOf.split('_')[1],
         }
       }
+      if (outcomeOccurrence.characteristicOccurrences)
+        for (const occ of outcomeOccurrence.characteristicOccurrences) {
+          if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleGeneric("outcomeOccurrence", outcomeOccurrence._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/outcomeOccurrence/OutcomeOccurrences.js:97]: Same as above

File: frontend/src/components/programProvision/programProvisions.js [hunk 5]
```diff
@@ -85,6 +81,9 @@ export default function ProgramProvisions() {
             appointmentData.startDate = occ.dataDateValue;
           } else if (occ.occurrenceOf?.name === 'End Date') {
             appointmentData.endDate = occ.dataDateValue;
+          } else if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleGeneric("programProvision", appointment._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/programProvision/programProvisions.js:85]: Same as above

File: frontend/src/components/programRegistration/ProgramRegistrations.js [hunk 6]
```diff
@@ -56,6 +48,9 @@ export default function ProgramRegistrations() {
             programData.referralType = occ.dataStringValue;
           } else if (occ.occurrenceOf?.name === 'Referral Status') {
             programData.referralStatus = occ.objectValue;
+          } else if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleGeneric("programRegistration", program._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/programRegistration/ProgramRegistrations.js:52]: Same as above

File: frontend/src/components/referrals/Referrals.js [hunk 7]
```diff
@@ -43,33 +44,32 @@ const columnsWithoutOptions = [
...
-          name: formatName(referral.client.firstName, referral.client.lastName),
+          name: formatName(referral.client.firstName, referral.client.lastName, 'client',  referral.client._id),
           id: referral.client._id
         }
-      });
+      };
+      if (referral.characteristicOccurrences)
+        for (const occ of referral.characteristicOccurrences) {
+          if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleGeneric("referral", referral._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/referrals/Referrals.js:68]: Same as above

File: frontend/src/components/serviceProvision/serviceProvisions.js [hunk 8]
```diff
@@ -85,6 +81,9 @@ export default function ServiceProvisions() {
             appointmentData.startDate = occ.dataDateValue;
           } else if (occ.occurrenceOf?.name === 'End Date') {
             appointmentData.endDate = occ.dataDateValue;
+          } else if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleGeneric("serviceProvision", appointment._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/serviceProvision/serviceProvisions.js:85]: Same as above

File: frontend/src/components/serviceRegistration/ServiceRegistrations.js [hunk 9]
```diff
@@ -56,6 +48,9 @@ export default function ServiceRegistrations() {
             serviceData.referralType = occ.dataStringValue;
           } else if (occ.occurrenceOf?.name === 'Referral Status') {
             serviceData.referralStatus = occ.objectValue;
+          } else if (occ.occurrenceOf?.name === 'Address') {
+            const obj = (await fetchSingleGeneric("serviceRegistration", service._id)).data; // TODO: inefficient!
```
Comment (LesterLyu) [inline frontend/src/components/serviceRegistration/ServiceRegistrations.js:52]: Same as above

File: frontend/src/components/shared/GenericPage.js [hunk 10]
```diff
@@ -113,9 +113,20 @@ export default function GenericPage(props) {
     ...props.tableOptions,
   };
 
+  const generateMarkers = (data, pageNumber, rowsPerPage) => {
+    const currPageItems = data.slice((pageNumber - 1) * rowsPerPage, pageNumber * rowsPerPage);
+    return currPageItems.map(item => ({
+      position: (item.address?.lat && item.address?.lng)
+        ? {lat: Number(item.address.lat), lng: Number(item.address.lng)}
+        : {...item.address},
+      title: nameFormatter(item),
```
Comment (LesterLyu) [inline frontend/src/components/shared/GenericPage.js:122]: We might have different format for different generic type, but this should be fine for now.

**Comment by LesterLyu** [comment]
We defined the properties to populate in each generic type [here](https://github.com/csse-uoft/SNM-I/blob/bcf85a70d2c6ae089c863c00e8f876a8ff71df59/backend/services/genericData/index.js#L133).

File: frontend/src/components/Providers.js (hunk 2 above)
Comment (htysc) [reply frontend/src/components/Providers.js]: I also added volunteer.address

**Comment by htysc** [comment]
https://github.com/csse-uoft/SNM-I/issues/91#issue-1958636671

**Review by LesterLyu — APPROVED** [review]

**Pull request merged**